mlc_llm/transform/fuse_split_rotary_embedding.py

```diff
@@ -10,237 +10,275 @@
 
-    @T.prim_func
+    Implementation is generic over the number of query heads,
+    key/value heads, sequence length, head dimension, and position
+    embedding base.  These parameters can be replaced with static
+    values using `PrimFunc.specialize`.
+    """
+
+    @T.prim_func(private=True)
     def split_rotary(
-        qkv: T.handle,
-        split_0: T.handle,
-        split_1: T.handle,
-        split_2: T.handle,
-        n: T.int64,
+        fused_qkv_handle: T.handle,
+        embedded_query_handle: T.handle,
+        embedded_key_handle: T.handle,
+        value_handle: T.handle,
+        rotary_offset: T.int64,
+        batch_size: T.int64,
+        seq_len: T.int64,
+        num_query_heads: T.int64,
+        num_kv_heads: T.int64,
+        head_dim: T.int64,
+        position_embedding_base: T.float32,
     ):
-        A = T.match_buffer(qkv, [1, 1, hidden_size * 3], dtype="float16")
-        T_split = T.match_buffer(split_0, [1, 1, hidden_size], dtype="float16")
-        T_split_1 = T.match_buffer(split_1, [1, 1, hidden_size], dtype="float16")
-        T_split_2 = T.match_buffer(split_2, [1, 1, hidden_size], dtype="float16")
+        Fused_QKV = T.match_buffer(
+            fused_qkv_handle,
+            [batch_size, seq_len, num_query_heads + num_kv_heads * 2, head_dim],
+            dtype="float16",
+        )
+        EmbeddedQuery = T.match_buffer(
+            embedded_query_handle,
+            [batch_size, seq_len, num_query_heads, head_dim],
+            dtype="float16",
+        )
+        EmbeddedKey = T.match_buffer(
+            embedded_key_handle,
+            [batch_size, seq_len, num_kv_heads, head_dim],
+            dtype="float16",
+        )
+        Value = T.match_buffer(
+            value_handle,
+            [batch_size, seq_len, num_kv_heads, head_dim],
+            dtype="float16",
+        )
 
         T.func_attr({"op_pattern": 2, "tir.noalias": T.bool(True)})
-        for ax0, ax1, ax2 in T.grid(T.int64(1), T.int64(1), T.int64(hidden_size)):
-            with T.block("T_split"):
-                v_ax0, v_ax1, v_ax2 = T.axis.remap("SSS", [ax0, ax1, ax2])
-                T.reads(
-                    A[v_ax0, v_ax1, v_ax2],
-                    A[v_ax0, v_ax1, v_ax2 + T.int64(hidden_size)],
-                    A[v_ax0, v_ax1, v_ax2 + T.int64(hidden_size * 2)],
-                )
-                T.writes(
-                    T_split[v_ax0, v_ax1, v_ax2],
-                    T_split_1[v_ax0, v_ax1, v_ax2],
-                    T_split_2[v_ax0, v_ax1, v_ax2],
-                )
-                pos: T.float32 = T.Cast("float32", n - T.int64(1))
+
+        for iters in T.grid(batch_size, seq_len, num_query_heads + num_kv_heads * 2, head_dim):
+            with T.block("FusedRotaryEmbeddingAndSplitQKV"):
+                batch_i, seq_i, head_num, head_i = T.axis.remap("SSSS", iters)
+                pos: T.float32 = T.Cast("float32", rotary_offset + seq_i - seq_len)
```

Comment:
What is `- seq_len` for?

Reply:
This is part of the rotary embedding definition, [here](https://github.com/mlc-ai/mlc-llm/blob/main/mlc_llm/relax_model/llama.py#L197).  (The `offset` passed in for `llama.py` is `kv_seq_len - query_seq_len`, while the `rotary_offset` passed in for the `split_rotary` is just `kv_seq_len`, because `query_seq_len` is inferred from the tensor shape.  I probably should rename the `rotary_offset` variable for it.)

For the decode step, the expression `rotary_offset + seq_i - seq_len` always had `seq_i == 0` and `seq_len == 1`, so this became `rotary_offset - 1`, but we need the full expression for the prefill step.